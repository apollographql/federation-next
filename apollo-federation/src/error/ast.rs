use apollo_compiler::ast::{
    Argument, Field, FragmentDefinition, FragmentSpread, InlineFragment, OperationDefinition,
    VariableDefinition,
};
use apollo_compiler::schema::{
    Directive, DirectiveDefinition, EnumType, EnumValueDefinition, FieldDefinition,
    InputObjectType, InputValueDefinition, InterfaceType, ObjectType, ScalarType, SchemaDefinition,
    Type, UnionType, Value,
};
use apollo_compiler::Node;

#[derive(Clone, Debug)]
pub enum AstNode {
    OperationDefinition(Node<OperationDefinition>),
    FragmentDefinition(Node<FragmentDefinition>),
    VariableDefinition(Node<VariableDefinition>),
    Field(Node<Field>),
    FragmentSpread(Node<FragmentSpread>),
    InlineFragment(Node<InlineFragment>),
    Directive(Node<Directive>),
    Argument(Node<Argument>),
    Type(Node<Type>),
    Value(Node<Value>),
    DirectiveDefinition(Node<DirectiveDefinition>),
    SchemaDefinition(Node<SchemaDefinition>),
    ScalarType(Node<ScalarType>),
    ObjectType(Node<ObjectType>),
    InterfaceType(Node<InterfaceType>),
    UnionType(Node<UnionType>),
    EnumType(Node<EnumType>),
    InputObjectType(Node<InputObjectType>),
    FieldDefinition(Node<FieldDefinition>),
    InputValueDefinition(Node<InputValueDefinition>),
    EnumValueDefinition(Node<EnumValueDefinition>),
}

impl From<Node<OperationDefinition>> for AstNode {
    fn from(value: Node<OperationDefinition>) -> Self {
        AstNode::OperationDefinition(value)
    }
}

impl From<Node<FragmentDefinition>> for AstNode {
    fn from(value: Node<FragmentDefinition>) -> Self {
        AstNode::FragmentDefinition(value)
    }
}

impl From<Node<VariableDefinition>> for AstNode {
    fn from(value: Node<VariableDefinition>) -> Self {
        AstNode::VariableDefinition(value)
    }
}

impl From<Node<Field>> for AstNode {
    fn from(value: Node<Field>) -> Self {
        AstNode::Field(value)
    }
}

impl From<Node<FragmentSpread>> for AstNode {
    fn from(value: Node<FragmentSpread>) -> Self {
        AstNode::FragmentSpread(value)
    }
}

impl From<Node<InlineFragment>> for AstNode {
    fn from(value: Node<InlineFragment>) -> Self {
        AstNode::InlineFragment(value)
    }
}

impl From<Node<Directive>> for AstNode {
    fn from(value: Node<Directive>) -> Self {
        AstNode::Directive(value)
    }
}

impl From<Node<Argument>> for AstNode {
    fn from(value: Node<Argument>) -> Self {
        AstNode::Argument(value)
    }
}

impl From<Node<Type>> for AstNode {
    fn from(value: Node<Type>) -> Self {
        AstNode::Type(value)
    }
}

impl From<Node<Value>> for AstNode {
    fn from(value: Node<Value>) -> Self {
        AstNode::Value(value)
    }
}

impl From<Node<DirectiveDefinition>> for AstNode {
    fn from(value: Node<DirectiveDefinition>) -> Self {
        AstNode::DirectiveDefinition(value)
    }
}

impl From<Node<SchemaDefinition>> for AstNode {
    fn from(value: Node<SchemaDefinition>) -> Self {
        AstNode::SchemaDefinition(value)
    }
}

impl From<Node<ScalarType>> for AstNode {
    fn from(value: Node<ScalarType>) -> Self {
        AstNode::ScalarType(value)
    }
}

impl From<Node<ObjectType>> for AstNode {
    fn from(value: Node<ObjectType>) -> Self {
        AstNode::ObjectType(value)
    }
}

impl From<Node<InterfaceType>> for AstNode {
    fn from(value: Node<InterfaceType>) -> Self {
        AstNode::InterfaceType(value)
    }
}

impl From<Node<UnionType>> for AstNode {
    fn from(value: Node<UnionType>) -> Self {
        AstNode::UnionType(value)
    }
}

impl From<Node<EnumType>> for AstNode {
    fn from(value: Node<EnumType>) -> Self {
        AstNode::EnumType(value)
    }
}

impl From<Node<InputObjectType>> for AstNode {
    fn from(value: Node<InputObjectType>) -> Self {
        AstNode::InputObjectType(value)
    }
}

impl From<Node<FieldDefinition>> for AstNode {
    fn from(value: Node<FieldDefinition>) -> Self {
        AstNode::FieldDefinition(value)
    }
}

impl From<Node<InputValueDefinition>> for AstNode {
    fn from(value: Node<InputValueDefinition>) -> Self {
        AstNode::InputValueDefinition(value)
    }
}

impl From<Node<EnumValueDefinition>> for AstNode {
    fn from(value: Node<EnumValueDefinition>) -> Self {
        AstNode::EnumValueDefinition(value)
    }
}
