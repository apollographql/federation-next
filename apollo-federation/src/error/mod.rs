pub mod ast;
pub mod error;
