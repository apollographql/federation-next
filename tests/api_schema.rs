use apollo_compiler::validation::Valid;
use apollo_compiler::Schema;
use apollo_federation::error::FederationError;
use apollo_federation::Supergraph;

const INACCESSIBLE_V02_HEADER: &str = r#"
    directive @link(url: String!, as: String, import: [link__Import], for: link__Purpose) repeatable on SCHEMA

    scalar link__Import

    enum link__Purpose {
      EXECUTION
      SECURITY
    }

    directive @inaccessible on FIELD_DEFINITION | OBJECT | INTERFACE | UNION | ARGUMENT_DEFINITION | SCALAR | ENUM | ENUM_VALUE | INPUT_OBJECT | INPUT_FIELD_DEFINITION

    schema
      @link(url: "https://specs.apollo.dev/link/v0.2")
      @link(url: "https://specs.apollo.dev/inaccessible/v0.2")
    {
      query: Query
    }
"#;

fn inaccessible_to_api_schema(input: &str) -> Result<Valid<Schema>, FederationError> {
    let sdl = format!("{INACCESSIBLE_V02_HEADER}{input}");
    let graph = Supergraph::new(&sdl).unwrap();
    graph.to_api_schema()
}

#[test]
fn inaccessible_types_with_accessible_references() {
    let errors = inaccessible_to_api_schema(
        r#"
      # Query types can't be inaccessible
      type Query @inaccessible {
        someField: String
      }

      # Inaccessible object type
      type Object @inaccessible {
        someField: String
      }

      # Inaccessible object type can't be referenced by object field in the API
      # schema
      type Referencer1 implements Referencer2 {
        someField: Object!
      }

      # Inaccessible object type can't be referenced by interface field in the
      # API schema
      interface Referencer2 {
        someField: Object
      }

      # Inaccessible object type can't be referenced by union member with a
      # non-inaccessible parent and no non-inaccessible siblings
      union Referencer3 = Object
    "#,
    )
    .expect_err("should return validation errors");

    insta::assert_display_snapshot!(errors);
}

#[test]
fn inaccessible_interface_with_accessible_references() {
    let errors = inaccessible_to_api_schema(
        r#"
      type Query {
        someField: String
      }

      # Inaccessible interface type
      interface Interface @inaccessible {
        someField: String
      }

      # Inaccessible interface type can't be referenced by object field in the
      # API schema
      type Referencer1 implements Referencer2 {
        someField: [Interface!]!
      }

      # Inaccessible interface type can't be referenced by interface field in
      # the API schema
      interface Referencer2 {
        someField: [Interface]
      }
    "#,
    )
    .expect_err("should return validation errors");

    insta::assert_display_snapshot!(errors);
}

#[test]
fn inaccessible_union_with_accessible_references() {
    let errors = inaccessible_to_api_schema(
        r#"
      type Query {
        someField: String
      }

      # Inaccessible union type
      union Union @inaccessible = Query

      # Inaccessible union type can't be referenced by object field in the API
      # schema
      type Referencer1 implements Referencer2 {
        someField: Union!
      }

      # Inaccessible union type can't be referenced by interface field in the
      # API schema
      interface Referencer2 {
        someField: Union
      }
    "#,
    )
    .expect_err("should return validation errors");

    insta::assert_display_snapshot!(errors);
}

#[test]
fn inaccessible_input_object_with_accessible_references() {
    let errors = inaccessible_to_api_schema(
        r#"
      type Query {
        someField: String
      }

      # Inaccessible input object type
      input InputObject @inaccessible {
        someField: String
      }

      # Inaccessible input object type can't be referenced by object field
      # argument in the API schema
      type Referencer1 implements Referencer2 {
        someField(someArg: InputObject): String
      }

      # Inaccessible input object type can't be referenced by interface field
      # argument in the API schema
      interface Referencer2 {
        someField(someArg: InputObject): String
      }

      # Inaccessible input object type can't be referenced by input object field
      # in the API schema
      input Referencer3 {
        someField: InputObject
      }

      # Inaccessible input object type can't be referenced by directive argument
      # in the API schema
      directive @referencer4(someArg: InputObject) on QUERY
    "#,
    )
    .expect_err("should return validation errors");

    insta::assert_display_snapshot!(errors);
}

#[test]
fn inaccessible_enum_with_accessible_references() {
    let errors = inaccessible_to_api_schema(
        r#"
      type Query {
        someField: String
      }

      # Inaccessible enum type
      enum Enum @inaccessible {
        SOME_VALUE
      }

      # Inaccessible enum type can't be referenced by object field in the API
      # schema
      type Referencer1 implements Referencer2 {
        somefield: [Enum!]!
      }

      # Inaccessible enum type can't be referenced by interface field in the API
      # schema
      interface Referencer2 {
        somefield: [Enum]
      }

      # Inaccessible enum type can't be referenced by object field argument in
      # the API schema
      type Referencer3 implements Referencer4 {
        someField(someArg: Enum): String
      }

      # Inaccessible enum type can't be referenced by interface field argument
      # in the API schema
      interface Referencer4 {
        someField(someArg: Enum): String
      }

      # Inaccessible enum type can't be referenced by input object field in the
      # API schema
      input Referencer5 {
        someField: Enum
      }

      # Inaccessible enum type can't be referenced by directive argument in the
      # API schema
      directive @referencer6(someArg: Enum) on FRAGMENT_SPREAD
    "#,
    )
    .expect_err("should return validation errors");

    insta::assert_display_snapshot!(errors);
}

#[test]
fn inaccessible_scalar_with_accessible_references() {
    let errors = inaccessible_to_api_schema(
        r#"
      type Query {
        someField: String
      }

      # Inaccessible scalar type
      scalar Scalar @inaccessible

      # Inaccessible scalar type can't be referenced by object field in the API
      # schema
      type Referencer1 implements Referencer2 {
        somefield: [[Scalar!]!]!
      }

      # Inaccessible scalar type can't be referenced by interface field in the
      # API schema
      interface Referencer2 {
        somefield: [[Scalar]]
      }

      # Inaccessible scalar type can't be referenced by object field argument in
      # the API schema
      type Referencer3 implements Referencer4 {
        someField(someArg: Scalar): String
      }

      # Inaccessible scalar type can't be referenced by interface field argument
      # in the API schema
      interface Referencer4 {
        someField(someArg: Scalar): String
      }

      # Inaccessible scalar type can't be referenced by input object field in
      # the API schema
      input Referencer5 {
        someField: Scalar
      }

      # Inaccessible scalar type can't be referenced by directive argument in
      # the API schema
      directive @referencer6(someArg: Scalar) on MUTATION
    "#,
    )
    .expect_err("should return validation errors");

    insta::assert_display_snapshot!(errors);
}

#[test]
fn remove_inaccessible() {
    // let s = api_schema(r#""#);
}
