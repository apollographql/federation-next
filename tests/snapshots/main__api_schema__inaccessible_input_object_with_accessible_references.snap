---
source: tests/api_schema.rs
expression: errors
---
The following errors occurred:

  - Type `InputObject` is @inaccessible but is referenced by `Referencer1.someField(someArg:)`, which is in the API schema.

  - Type `InputObject` is @inaccessible but is referenced by `Referencer2.someField(someArg:)`, which is in the API schema.

  - Type `InputObject` is @inaccessible but is referenced by `Referencer3.someField`, which is in the API schema.

  - Type `InputObject` is @inaccessible but is referenced by `@referencer4(someArg:)`, which is in the API schema.
